2020/webkitgtk-npapi-sunsetting.markdown

```diff
@@ -0,0 +1,195 @@
+  <li><a href="#what-are-other-browsers-doing">What are other browsers doing?</a></li>
+  <li><a href="#is-webkitgtk-following-suit">Is WebKitGTK following suit?</a></li>
+</ol>
+
+## Summary
+
+Here's a [tl;dr](https://www.urbandictionary.com/define.php?term=tl%3Bdr)
+list of bullet points:
+
+* One year ago, [WebKitGTK 2.26.0](https://www.webkitgtk.org/2019/09/09/webkitgtk2.26.0-released.html)
+  [removed
+  support](/2019/webkit-2.26-six-month-retrospective.html#cleanups) for NPAPI
+  plug-ins which used GTK2, but the rest of plug-ins kept working.
+* WebKitGTK 2.30.x will be the last stable series with support for NPAPI
+  plug-ins _at all_. Version [2.30.0 was
+  released](https://webkitgtk.org/2020/09/11/webkitgtk2.30.0-released.html) a
+  couple of weeks ago.
+* WebKitGTK 2.32.0, due in March 2021, will be the first *stable* release
+  to ship without support for NPAPI plug-ins.
+* We have already [removed the relevant
+  code](https://trac.webkit.org/changeset/265753/) from the WebKit repository.
+* While the WPE WebKit port allowed running windowless NPAPI plug-ins, this
+  was never advertised nor supported by us.
+
+
+## What is NPAPI?
+
+NPAPI, short for _Netscape Plugin Application Programming Interface_ allowed
+browser plug-ins to extend the functionality of a web browser providing
+support for additional content types. NPAPI appeared first in Netscape
+Navigator 2.0, in 1995.
+
+When a NPAPI plug-in is used to render content, the web browser will
+effectively _carve a hole_ in the rectangular location where content handled
+by the plug-in will be placed, and hands off the rendering responsibility to
+the plug-in.
+
+## What is NPAPI used for?
+
+A number of technologies have used NPAPI along the years for different
+purposes:
+
+* Displaying of multimedia content using [Flash Player][wp-flash] or the
+  [Silverlight][wp-silverlight] plug-ins.
+* Running rich [Java™][wp-javaplug] applications in the browser.
+* Displaying documents in non-Web formats (PDF, [DjVu][wp-djvu]) inside
+  browser windows.
+* A number of questionable practices, like <abbr title="Virtual Private
+  Network">VPN</abbr> client software using a browser plug‑in for
+  configuration.
+
+[wp-silverlight]: https://en.wikipedia.org/wiki/Microsoft_Silverlight
+[wp-flash]: https://en.wikipedia.org/wiki/Adobe_Flash_Player
+[wp-javaplug]: https://en.wikipedia.org/wiki/Java_virtual_machine#JVM_in_the_web_browser
+[wp-djvu]: https://en.wikipedia.org/wiki/DjVu#Role_in_the_software_ecosystem
+
+
+## Why are NPAPI plug-ins being phased out?
+
+Due to the way this was done means that the browser is no longer in control
+of how the content is displayed. This not only makes it harder to style and
+do layout for them, but also it means that the browser has no control
+whatsoever over what plug-ins do—and they can do lots, given they are
+compiled, native code—, which historically has resulted in plug-ins being
+the cause for security incidents, crashes, and browser hangs.
```

Comment:
It is good to avoid passive voice (passive voice is not how we speak). Also the secondary
text breaks the reading (the one between em-dash signs). Because it is easier said than
done, I would suggest something like:

> The NPAPI design makes the web browser gives full control of the content to the plug-in.
> It is hard to style and do layout for them. More importantly, the browser has no control
> whatsoever over what plug-ins do, and the result is a history of security incidents, crashes,
> and browser hangs. This is because the plug-in can do a lot, given they are compiled,
> native code.

Perhaps, the following idea is missing:

_All the effort that browser developers put on quality control to secure the web browser is
handed over to a plug-in._

Reply:
Agh, I wrote such a mouthful here! This is a hard paragraph to get right. I gave your version another spin, trying to add the idea that browser developers cannot do QA on plug-ins, and ended up with this:

> The design of NPAPI  makes the web browser give full responsibility to plug-ins: the browser has no control whatsoever over what plug-ins do to display content, which makes it hard to make them participate in styling and layout. More importantly, plug-ins are compiled, native code over which browser developers cannot exercise quality control, which resulted in a history of security incidents, crashes, and browser hangs.

Reply:
Much better.